### PR TITLE
Add format as string or number

### LINKED ISSUE #7505: Nicely format strings in Yul IR generation
When generating the Yul intermediate code of Solidity programs, strings are always translated to hex numbers. This prevents readability of IR code. Because of that, the code generator should detect if a string can be represented in printable ascii chars (for safety) and use that instead.

## How to implement:

Add a function `std::string formatAsStringOrNumber(std::string const& _value)` in `libdevcore/CommonData.h` that checks if all characters are safe and outputs a quoted string. If there is an unsafe charater, use `"0x" + h256(_value, h256::AlignLeft).hex();` as it has been used before (you need to include `FixedHash.h` for that).

Now use that function in ABIFunctions.cpp:978 (`ABIFunctions::abiEncodingFunctionStringLiteral`) and :993. The same change can be applied to `YulUtilFunctions::conversionFunctionSpecial`.

## How to test:

Run `solc --ir` on e.g.
```
contract c {
function f() external pure returns (string memory) { return "abcabc"; }
}
```

It should not contain `0x6162636162630000000000000000000000000000000000000000000000000000` anymore but `"abcabc"` instead.

## How to add an automated test

Create a copy of the directory `test/cmdlineTests/standard_ir_requested` such that `input.json` contains code similar to the one above. Run `./test/cmdlineTests.sh` to update the expectation file.

You should additionally create a test inside `test/libsolidity/semanticTests` that ends in

```
// ====
// compileViaYul: also
// ----
// f()
```

This test should check that the contract still works as expected. Please also try string literals longer than 32 bytes.

obsoletes https://github.com/ethereum/solidity/issues/7495

### PR DESCRIPTION
**Pull request by secondW4ve**
Fixes https://github.com/ethereum/solidity/issues/7505

### Checklist
- [ ] Code compiles correctly
- [ ] All tests are passing
- [ ] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary
- [ ] Changelog entry (if change is visible to the user)
- [ ] Used meaningful commit messages

### CODE REVIEW

**Comment by secondW4ve** [comment]
I need help with testing what I have written and adding an automated test, and ofc with another mistakes. Dont be mad, its my first pull request

File: libsolidity/codegen/ABIFunctions.cpp [hunk 1]
```diff
@@ -975,7 +975,7 @@ string ABIFunctions::abiEncodingFunctionStringLiteral(
 			for (size_t i = 0; i < words; ++i)
 			{
 				wordParams[i]["offset"] = to_string(i * 32);
-				wordParams[i]["wordValue"] = "0x" + h256(value.substr(32 * i, 32), h256::AlignLeft).hex();
+				wordParams[i]["wordValue"] = "0x" + dev::formatAsStringOrNumber(value.substr(32 * i, 32);
```
Comment (chriseth) [inline libsolidity/codegen/ABIFunctions.cpp]: You have to remove the `"0x" +` in all three calls.

File: libdevcore/CommonData.h [hunk 2]
```diff
@@ -394,4 +393,23 @@ std::vector<T> make_vector(Args&&... _args)
 	return result;
 }
 
+// Function that check if all characters in string are safe and return quoted parametr.
+// If there is unsafe characters it use "0x" + h256(_value, h256::AlignLeft).hex();
+std::string formatAsStringOrNumber(std::string const &_value)
```
Comment (chriseth) [inline libdevcore/CommonData.h]:
```suggestion
std::string formatAsStringOrNumber(std::string const& _value)
```

File: libdevcore/CommonData.h [hunk 3]
```diff
@@ -394,4 +393,23 @@ std::vector<T> make_vector(Args&&... _args)
...
+{
+	std::string resultString = "\"";
+	for (int i = 0; i < _value.length; i++)
```
Comment (chriseth) [inline libdevcore/CommonData.h]: You can use a range-based for loop here.

File: libdevcore/CommonData.h [hunk 4]
```diff
@@ -394,4 +393,23 @@ std::vector<T> make_vector(Args&&... _args)
...
+	{
+		if (!isalpha(_value[i]))
+		{
+			return "0x" + h256(_value, h256::AlignLeft).hex();
+		}
+		else
+		{
+			resultString += _value[i];
```
Comment (chriseth) [inline libdevcore/CommonData.h]: No need to copy each character individually, you can just return the full string (prefix and postfixed with quotes) at the ond.

File: libdevcore/CommonData.h [hunk 5]
```diff
@@ -394,4 +393,23 @@ std::vector<T> make_vector(Args&&... _args)
...
 }
 
+// Function that check if all characters in string are safe and return quoted parametr.
+// If there is unsafe characters it use "0x" + h256(_value, h256::AlignLeft).hex();
+std::string formatAsStringOrNumber(std::string const &_value)
+{
+	std::string resultString = "\"";
+	for (int i = 0; i < _value.length; i++)
+	{
+		if (!isalpha(_value[i]))
```
Comment (chriseth) [inline libdevcore/CommonData.h]: I think `isalpha` is a bit too restrictive. Also, can it depend on the locale?

**Comment by chriseth** [comment]
No worries, we are here to help! Were you able to test it manually as described in the issue?

File: libdevcore/CommonData.h (hunk 5 above)
Comment (secondW4ve) [reply libdevcore/CommonData.h]: Yea, I think you are right. What should I use there then?

**Comment by secondW4ve** [comment]
> 
> 
> No worries, we are here to help! Were you able to test it manually as described in the issue?

No, I am not familiar with testing. Where should I run solc --ir? Do I have to build all project somehow, etc?

**Comment by chriseth** [comment]
Yes, you have to build everything, then put this in a file called `x.sol`:
```
contract c {
function f() external pure returns (string memory) { return "abcabc"; }
}
```
and run `solc --ir x.sol` and check the output. Of course `solc` either has to be in your path or you specify the full path to the binary in a subdirectory of the `build` directory.

File: libdevcore/CommonData.h [hunk 6]
```diff
@@ -394,4 +393,20 @@ std::vector<T> make_vector(Args&&... _args)
...
 }
 
+// Function that check if all characters in string are safe and return quoted parameter.
+// If there is unsafe characters it use "0x" + h256(_value, h256::AlignLeft).hex();
+std::string formatAsStringOrNumber(std::string const& _value)
+{
+	std::string resultString = "\"";
+	for (auto i:value)
+	//for (int i = 0; i < _value.length; i++)
+	{
```
Comment (bshastry) [inline libdevcore/CommonData.h:403]: braces not needed since the for loop contains a single statement

File: libdevcore/CommonData.h [hunk 7]
```diff
@@ -394,4 +393,20 @@ std::vector<T> make_vector(Args&&... _args)
...
+		if (!isalpha(i))
+		{
```
Comment (bshastry) [inline libdevcore/CommonData.h:405]: same here

File: libdevcore/CommonData.h [hunk 8]
```diff
@@ -394,4 +393,20 @@ std::vector<T> make_vector(Args&&... _args)
 	return result;
 }
 
+// Function that check if all characters in string are safe and return quoted parameter.
```
Comment (bshastry) [inline libdevcore/CommonData.h:396]: Please use three forward slashes `///` for documenting function spec.

Comment (bshastry) [reply libdevcore/CommonData.h:396]: Minor typo: `Function that checks... and returns quoted parameter.`

File: libdevcore/CommonData.h [hunk 9]
```diff
@@ -394,4 +393,20 @@ std::vector<T> make_vector(Args&&... _args)
...
+// If there is unsafe characters it use "0x" + h256(_value, h256::AlignLeft).hex();
```
Comment (bshastry) [inline libdevcore/CommonData.h:397]: `If there are unsafe characters it uses a "0x" prefix`?

File: libdevcore/CommonData.h [hunk 10]
```diff
@@ -394,4 +393,20 @@ std::vector<T> make_vector(Args&&... _args)
...
+std::string formatAsStringOrNumber(std::string const& _value)
+{
+	std::string resultString = "\"";
+	for (auto i:value)
```
Comment (bshastry) [inline libdevcore/CommonData.h:401]: Please add a whitespace between colon and `value`

File: libdevcore/CommonData.h [hunk 11]
```diff
@@ -1,16 +1,13 @@
 /*
 	This file is part of solidity.
-
```
Comment (bshastry) [inline libdevcore/CommonData.h:3]: Please keep the text header intact.

File: libdevcore/CommonData.h [hunk 12]
```diff
@@ -294,6 +291,8 @@ void iterateReplacing(std::vector<T>& _vector, F const& _f)
 }
 
 
+
```
Comment (bshastry) [inline libdevcore/CommonData.h:294]: Too many newlines here

**Review by bshastry — CHANGES_REQUESTED** [review]
@secondW4ve please let us know if you are stuck and need help with something. In the meantime, I added a few more review comments.

**Comment by chriseth** [comment]
@secondW4ve do you still plan to work on this?

**Comment by Marenz** [comment]
@secondW4ve still waiting for a reply from you here :)

**Comment by leonardoalt** [comment]
I took over this PR and finished it, but apparently not allowed to pushed to the fork. So I'll open another PR.

**Comment by leonardoalt** [comment]
Other PR: https://github.com/ethereum/solidity/pull/7770

**Comment by Marenz** [comment]
Closing in favor of #7770

**Pull request closed without merging**